kite_provider/provider.py

```diff
@@ -0,0 +1,229 @@
+        self.available_languages = []
+        self.client = KiteClient(None)
+        self.kite_process = None
+
+        # Signals
+        self.client.sig_client_started.connect(self.http_client_ready)
+        self.client.sig_status_response_ready[str].connect(
+            self.set_status)
+        self.client.sig_status_response_ready[dict].connect(
+            self.set_status)
+        self.client.sig_response_ready.connect(
+            functools.partial(self.sig_response_ready.emit,
+                              self.COMPLETION_PROVIDER_NAME))
+
+        self.client.sig_client_wrong_response.connect(
+            self._wrong_response_error)
+
+        # Status bar widget
+        self.STATUS_BAR_CLASSES = [
+            self.create_statusbar
+        ]
+
+        # Config
+        self.update_kite_configuration(self.config)
+
+
+    # ------------------ SpyderCompletionProvider methods ---------------------
+    def get_name(self):
+        return 'Kite'
+
+    def send_request(self, language, req_type, req, req_id):
+        if language in self.available_languages:
+            self.client.sig_perform_request.emit(req_id, req_type, req)
+        else:
+            self.sig_response_ready.emit(self.COMPLETION_PROVIDER_NAME,
+                                         req_id, {})
+
+    def start_completion_services_for_language(self, language):
+        return language in self.available_languages
+
+    def start(self):
+        try:
+            installed, path = check_if_kite_installed()
+            if not installed:
+                return
+            logger.debug('Kite was found on the system: {0}'.format(path))
+            running = check_if_kite_running()
+            if running:
+                return
+            logger.debug('Starting Kite service...')
+            self.kite_process = run_program(path)
+        except OSError:
+            installed, path = check_if_kite_installed()
+            logger.debug(
+                'Error starting Kite service at {path}...'.format(path=path))
+            if self.get_conf('show_installation_error_message'):
+                err_str = _(
+                    "It seems that your Kite installation is faulty. "
+                    "If you want to use Kite, please remove the "
+                    "directory that appears bellow, "
+                    "and try a reinstallation:<br><br>"
+                    "<code>{kite_dir}</code>").format(
+                        kite_dir=osp.dirname(path))
+
+        finally:
+            # Always start client to support possibly undetected Kite builds
+            self.client.start()
+
+    def shutdown(self):
+        try:
+            self.remove_item_from_application_menu(
+                KiteProviderActions.Installation,
+                menu_id=ApplicationMenus.Tools)
+        except KeyError:
+            # Action does not exist
+            pass
+
```

Comment:
We should be able now to remove this too (no action to install Kite should be added to the main menu):

```suggestion
```